scapy/contrib/pnio_rpc.py

```diff
@@ -0,0 +1,1003 @@
+        BitEnumField("SubmoduleProperties_Type", 0, 2,
+                     ["NO_IO", "INPUT", "OUTPUT", "INPUT_OUTPUT"]),
+        PacketListField(
+            "DataDescription", [], ExpectedSubmoduleDataDescription,
+            count_from=lambda p: 2 if p.SubmoduleProperties_Type == 3 else 1
+        ),
+    ]
+
+    def extract_padding(self, s):
+        return None, s  # No extra payload
+
+
+class ExpectedSubmoduleAPI(Packet):
+    """Description of an API in the expected submodules blocks"""
+    name = "API"
+    fields_desc = [
+        XIntField("API", 0),
+        XShortField("SlotNumber", 0),
+        XIntField("ModuleIdentNumber", 0),
+        XShortField("ModuleProperties", 0),
+        FieldLenField("NumberOfSubmodules", None, fmt="H",
+                      count_of="Submodules"),
+        PacketListField("Submodules", [], ExpectedSubmodule,
+                        count_from=lambda p: p.NumberOfSubmodules),
+    ]
+
+    def extract_padding(self, s):
+        return None, s  # No extra payload
+
+
+class ExpectedSubmoduleBlockReq(Block):
+    """Expected submodule block request"""
+    fields_desc = [
+        BlockHeader,
+        FieldLenField("NumberOfAPIs", None, fmt="H", count_of="APIs"),
+        PacketListField("APIs", [], ExpectedSubmoduleAPI,
+                        count_from=lambda p: p.NumberOfAPIs)
+    ]
+    # default block_type value
+    block_type = 0x0104
+
+    def get_response(self):
+        """Generate the response block of this request.
+        Carefull: it only sets the fields which can be set from the request
+        """
+        return None  # no response associated (should be modulediffblock)
+
+
+# PROFINET IO DCE/RPC PDU
+PNIO_RPC_BLOCK_ASSOCIATION = {
+    # requests
+    "0101": ARBlockReq,
+    "0102": IOCRBlockReq,
+    "0104": ExpectedSubmoduleBlockReq,
+    "0110": IODControlReq,
+    "0111": IODControlReq,
+    "0112": IODControlReq,
+    "0113": IODControlReq,
+    "0114": IODControlReq,
+    "0116": IODControlReq,
+    "0117": IODControlReq,
+    "0118": IODControlReq,
+
+    # responses
+    "8101": ARBlockRes,
+    "8102": IOCRBlockRes,
+    "8110": IODControlRes,
+    "8111": IODControlRes,
+    "8112": IODControlRes,
+    "8113": IODControlRes,
+    "8114": IODControlRes,
+    "8116": IODControlRes,
+    "8117": IODControlRes,
+    "8118": IODControlRes,
+}
+
+
+def _guess_block_class(_pkt, *args, **kargs):
+    cls = Block  # Default block type
+
+    # Special cases
+    if _pkt[:2] == b'\x00\x08':  # IODWriteReq
+        if _pkt[34:36] == b'\xe0@':  # IODWriteMultipleReq
+            cls = IODWriteMultipleReq
+        else:
+            cls = IODWriteReq
+
+    elif _pkt[:2] == b'\x80\x08':    # IODWriteRes
+        if _pkt[34:36] == b'\xe0@':  # IODWriteMultipleRes
+            cls = IODWriteMultipleRes
+        else:
+            cls = IODWriteRes
+
+    # Common cases
+    else:
+        btype = bytes_hex(_pkt[:2])
+        if btype in PNIO_RPC_BLOCK_ASSOCIATION:
```

Comment:
I think this doesn't work, because you compare a byte-string with a string.

I think you have to use `bytes_hex(b"\x01\x01").decode("utf-8")
` instead of `bytes_hex(b"\x01\x01")` in line 895

Reply:
Indeed it was a mistake, thanks for the review!